flink-runtime/src/test/java/org/apache/flink/runtime/rest/FileUploadHandlerTest.java

```diff
@@ -0,0 +1,471 @@
+		when(mockRestfulGateway.requestRestAddress(any(Time.class))).thenReturn(CompletableFuture.completedFuture(restAddress));
+
+		final GatewayRetriever<RestfulGateway> mockGatewayRetriever = () ->
+			CompletableFuture.completedFuture(mockRestfulGateway);
+
+		file1 = TEMPORARY_FOLDER.newFile();
+		Files.write(file1.toPath(), "hello".getBytes(ConfigConstants.DEFAULT_CHARSET));
+		file2 = TEMPORARY_FOLDER.newFile();
+		Files.write(file2.toPath(), "world".getBytes(ConfigConstants.DEFAULT_CHARSET));
+
+		mixedHandler = new MultipartMixedHandler(CompletableFuture.completedFuture(restAddress), mockGatewayRetriever);
+		jsonHandler = new MultipartJsonHandler(CompletableFuture.completedFuture(restAddress), mockGatewayRetriever);
+		fileHandler = new MultipartFileHandler(CompletableFuture.completedFuture(restAddress), mockGatewayRetriever);
+
+		final List<Tuple2<RestHandlerSpecification, ChannelInboundHandler>> handlers = Arrays.asList(
+			Tuple2.of(mixedHandler.getMessageHeaders(), mixedHandler),
+			Tuple2.of(jsonHandler.getMessageHeaders(), jsonHandler),
+			Tuple2.of(fileHandler.getMessageHeaders(), fileHandler));
+
+		serverEndpoint = new TestRestServerEndpoint(serverConfig, handlers);
+
+		serverEndpoint.start();
+		serverAddress = serverEndpoint.getRestBaseUrl();
+	}
+
+	@AfterClass
+	public static void teardown() throws Exception {
+		if (serverEndpoint != null) {
+			serverEndpoint.close();
+			serverEndpoint = null;
+		}
+	}
+
+	private static Request buildRequest(String headerUrl, int index, boolean includeFile, boolean includeJson) throws IOException {
+		Preconditions.checkArgument(includeFile || includeJson, "You have to either include JSON or a file.");
+		MultipartBody.Builder builder = new MultipartBody.Builder();
+
+		if (includeFile) {
+			okhttp3.RequestBody filePayload1 = okhttp3.RequestBody.create(MediaType.parse("application/octet-stream"), file1);
+			builder = builder.addFormDataPart("file1", file1.getName(), filePayload1);
+
+			okhttp3.RequestBody filePayload2 = okhttp3.RequestBody.create(MediaType.parse("application/octet-stream"), file2);
+			builder = builder.addFormDataPart("file2", file2.getName(), filePayload2);
+		}
+
+		if (includeJson) {
+			TestRequestBody jsonRequestBody = new TestRequestBody(index);
+
+			StringWriter sw = new StringWriter();
+			OBJECT_MAPPER.writeValue(sw, jsonRequestBody);
+
+			String jsonPayload = sw.toString();
+
+			builder = builder.addFormDataPart(org.apache.flink.runtime.rest.FileUploadHandler.HTTP_ATTRIBUTE_REQUEST, jsonPayload);
+		}
+
+		MultipartBody multipartBody = builder
+			.setType(MultipartBody.FORM)
+			.build();
+
+		return new Request.Builder()
+			.url(serverAddress + headerUrl)
+			.post(multipartBody)
+			.build();
+	}
+
+	@Test
+	public void testMixedMultipart() throws Exception {
+		OkHttpClient client = new OkHttpClient();
+
+		Request jsonRequest = buildRequest(mixedHandler.getMessageHeaders().getTargetRestEndpointURL(), RANDOM.nextInt(), false, true);
```

Comment:
Would be good to have methods `buildFileRequest` and `buildJsonRequest` and `buildJsonFileRequest` instead of the version with boolean. Booleans make it harder to understand which request is generated.